### DISCUSSION TITLE
Differential lines with lumped ports

### DISCUSSION THREAD

**Discussion by nico-arnold** [thliebig/openEMS-Project#180] [Q&A Usage]
Hey everybody!

I'm going to simulate a more complex microstrip structure with multiple layers and vias. However, the results seemed a bit too small for me as I'm opting for a 100 ohms differential impedance. To start small again now, I began constructing a piece of coupled microstrip line to start off getting the excitation and impedance plot right. However, it seems that I failed already there, as I get results around 55 ohms, while the calculator tool TX-Line tells me I should have around 165.

For a start, it seems like I already got the excitation correct (sort of); Current is travelling along the traces in opposing directions.
![image](https://github.com/thliebig/openEMS-Project/assets/82146516/2e2e16e4-e7f0-49f8-8f92-b32dc8041c06)

However, I still get values that are far too small when simulating against PMLs at the start and end of the line:
![image](https://github.com/thliebig/openEMS-Project/assets/82146516/19a203cd-c41d-4007-bb5b-d90350d57d20)

Even pushing the ports further inside the model and increasing the resolution from Lambda/50 to higher values and finally even Lambda/400 only made the impedance a small bit faster (and brought me to 10+ times the simulation time):
![image](https://github.com/thliebig/openEMS-Project/assets/82146516/cfd4e186-c9a3-4586-abdb-7e6f74e45bf0)

When I configure the simulation for using resistive ports and don't let the traces end in the PML, it looks like this:
![image](https://github.com/thliebig/openEMS-Project/assets/82146516/307b53e4-e103-45c9-9a2d-4a01a4a124f4)
...which would make more sense to me. Big mismatch in impedance, so we have a periodic imaginary overlay on the impedance plot.

The model is mostly based on the CPW_Line example, but with the ports changed to Lumped Ports instead of the CPW Ports. Maybe that's the problem. But I don't know if I can easily work with the CPW Ports here, as the single traces will be on different layers of the PCB later.
Is it necessary to excite the single traces against ground? In theory, for a symmetrical design, there should be a 0-Volt-Potential on GND all the time.

[You can find the model here](https://datashare.tu-dresden.de/s/YH7KJRk9Q9gKkD2). For the 2nd plot, just change line 26 to `if 1`.

thanks a lot in advance and have a great rest of the week,
Nico

**Comment by VolkerMuehlhaus** [comment]
@nico-arnold 

Hi Nico, I think the setup with **lumped ports** at some distance from the boundary makes no sense if the line extends beyond the ports to the boundary.  In that case, excitation and measurement plane are placed somewhere "inside" the line, and the signal propagates to left and right from the port.

I would draw the line between the ports only, and use the port resistors for termination. This is one on your testcases, can you post that model here so that we can have a look?

**Reply by VolkerMuehlhaus** [reply]
@nico-arnold Sorry for reading your post in a hurry, you mentioned the code change required to draw the line between the ports only, and used port termination resistors. 

The differential line impedance that I get from 2D cross section EM of your line, with zero metal thickness, is 152 Ohm. This is what I wanted to set as openEMS port impedance in the model, instead of 50 Ohm in your model, but my openEMS installation doesn't like Octave any more, and dies during simulation.

Using the (wrong) 50 Ohm for port impedance, the results that you see look somewhat reasonable. Below is the result that I get from circuit simulation using 50 Ohm ports for your 152 Ohm line.:

![diffline_port50](https://github.com/thliebig/openEMS-Project/assets/47922981/d73bc0e2-e82d-4767-beb4-7e3101d6bbc4)

And this is what you get when using 152 Ohm port impedance instead:

![diffline_port152](https://github.com/thliebig/openEMS-Project/assets/47922981/02f3c5fb-939e-48ea-98e5-cd0b98902401)

**Comment by biergaizi** [comment]
If you have doubts about using a differential excitation directly, you can check the results against the following brute-force solution to measure differential pairs with single-ended excitation - which is guaranteed to work. It's how you would do the job with a two-port VNA.

1. Excite port 1 only, find S11, S21, S31, S41.
2. Rerun simulation, excite port 2 only, find S22, S12, S32, S42.
3. Rerun simulation, excite port 3 only, find S33, S13, S23, S43.
4. Rerun simulation, excite port 4 only, find S44, S14, S24, S34.
5. Convert single-ended S-parameters to mixed-mode S-parameters, for example using [scikit-rf](https://scikit-rf.readthedocs.io/en/latest/examples/mixedmodeanalysis/Mixed%20Mode%20Basics.html).

Update: Since a differential pair is both symmetric and reciprocal, you only need to do half of those measurements and manually fill in the gaps.

**Reply by VolkerMuehlhaus** [reply]
> Is it necessary to excite the single traces against ground? In theory, for a symmetrical design, there should be a 0-Volt-Potential on GND all the time.

Let me add two more details on this topic:

To **check if the lumped excitation really creates differential voltages** on the line with respect to ground, you can use a differential lumped port for the excited source, and two lumped ports to ground on the other end. Then you can check S21 and S31 amplitude and phase, and see if they are indeed same magnitude with 180° phase offset.

But there is more detail: If you use lumped ports in differential configuration, this **will only terminate the **differential** mode**, and any common mode signal will see an open end. This is not an issue for the straight line in your testcase, but a real layout with some curves etc. might cause mode conversion. When working on such models with unterminated common mode, I had seen funny resonances, which disappeared when the common mode was terminated. 

One way to **terminate both differential mode and common mode** is the 4-port configuration mentioned by biergaizi above. For 100 Ohm differential line impedance, you would then use two 50 Ohm ports on each end, and get 50+50=100 Ohm for differential port impedance and 50||50=25 Ohm common mode port impedance. The downside of that external processing in scikit-rf is that you need to simulate all 4 port excitation one after another to get the full [S] matrix, before you can then extract the differential mode data.

**Comment by nico-arnold** [comment]
@VolkerMuehlhaus Indeed, the impedance mismatch was the reason again. When I change the value to 132 ohms (in my case), I get a nearly straight line. 

![image](https://github.com/thliebig/openEMS-Project/assets/82146516/b56bf6a9-3b2c-4014-b7b5-47f09cb2b83b)
Much lower than your calculated value. But that's because I'm using lossy lines here now (:
After the last  thread #146 , I somehow suspected I would see the mismatch in strange results, but somehow I might have not taken into account that I had PMLs at both ends... So in the end, if I get it right, I get a mismatch simulated (like we see in the resistor-terminated case), but I don't see any problem from the occuring reflections for example because the line is terminated by a PML swallowing everything? I expected at least the impedance would be in the right ballpark.

Is that Keysight ADS you're using there, BTW? 

@biergaizi Ah, good idea. Didn't think of that. It would be best if everything was possible within the "normal", but for automating that later from the PCB export it would definitely be a great option. Especially with the hint from @VolkerMuehlhaus that this terminates the common mode correctly.
What would be another option to do that? Using boundary conditions to force only certain modes on the line?

Meanwhile, I'm playing with some vias on the line. I changed over the lines and the ground layer to the other side of the substrate midway, and after adjusting the mesh I get my waves to pass down to the other side of the PCB:
![image](https://github.com/thliebig/openEMS-Project/assets/82146516/5739dd5b-2681-4325-ac45-b6251dc626b2)

![image](https://github.com/thliebig/openEMS-Project/assets/82146516/905148e3-a29f-4136-bcf5-cc406b08c99b)
(I used the old STL file to plot on here, that's why the GND vias are missing)

Impedance would also appear to make sense to me. Plot with 132 ohm feed:
![image](https://github.com/thliebig/openEMS-Project/assets/82146516/4666aaa7-034e-4dc4-9b6f-e0ff078313d3)

Just a small change on the feed impedance makes the S11 dip instantly:
![image](https://github.com/thliebig/openEMS-Project/assets/82146516/c6a5d3f2-8c9d-4b3c-b3ad-d9c9629b1ad3)

When I correct it upwards, there's a similar effect. Actually, there now should not be a "sweet spot" if I'm not mistaken - is that correct? We don't have a homogenous line as before, but different impedances one after another, the waves could get reflected on the top/bottom transition. So the next point for that example would be to adjust the via diameter and the ground spacing on the bottom of the PCB if one would prefer a perfect match.

[The model can be found here.](https://datashare.tu-dresden.de/s/8nCT43MFHk7kRio)

**Reply by VolkerMuehlhaus** [reply]
> but I don't see any problem from the occuring reflections for example because the line is terminated by a PML swallowing everything? I expected at least the impedance would be in the right ballpark.

Yes, you don't see a reflection because the PML absorbs everything. But still, the result is strange because your resulting "port in the middle of the line" approach excites a wave to **both** directions, so you measure the combined result of **both** directions. I suspect that you will get Zline/2 because from the excitation viewpoint, it is now measuring two lines (one in each direction) in parallel. 

> Is that Keysight ADS you're using there, BTW?

Yes, that screenshot is from the ADS circuit model. For impedance calculation I used the 2D cross section solver ("Controlled Impedance Line Designer") that works based on the EM substrate definition, because I wasn't sure how accurate the equation-based tools (TXLine or Linecalc) are for your dimensions.

In the meantime, I fixed my issue with Octave by installing the latest version, and was now able to run your line with the modification suggested above, two ports to ground that terminate the line and allow to check amplitude and phase difference. Result: with excitation from a differential lumped port, we indeed have equal amplitude and 180° phase offset on the two lines, when measured with respect to ground below. So that part worked as expected.

Regarding your latest model with the via, I recommend to have a look at this thread also:
https://github.com/thliebig/openEMS-Project/discussions/177
Looking at discontinuities in the time domain is a great tool to optimize such transitions, because you can "see" at what location along the line you are too low/high impedance or in other words, you L/C ratio isn't correct. I use that a lot for optimizing such transitions.

**Reply by nico-arnold** [reply]
> I suspect that you will get Zline/2 because from the excitation viewpoint, it is now measuring two lines (one in each direction) in parallel.

Hmm, good point indeed. Would make sense. But can that happen even when the line is terminated by the PML before the wave heads back to the port? I mean, isn't that part "invisible" to the port with the wave being absorbed?

>  2D cross section solver ("Controlled Impedance Line Designer") that works based on the EM substrate definition

very interesting. I think I saw something similar in Open Source a few weeks ago, but I think it is not maintained anymore. Would surely be a great alternative and also a good way to verify results for linear structures.

> In the meantime, I fixed my issue with Octave 

Ah, good to hear that. So we can assume the path we're taking here for excitation and measuring works and will give good results in most cases?
I will check the thread tomorrow morning, thanks a lot for the lecture tip! Looking at the time domain is indeed a technique I didn't think of yet. But it appears reasonable to me. Good idea!

best wishes and have a good start into the week tomorrow,
Nico

**Reply by VolkerMuehlhaus** [reply]
> But can that happen even when the line is terminated by the PML before the wave heads back to the port? I mean, isn't that part "invisible" to the port with the wave being absorbed?

Yes, with PML termination the port will "see" the line impedance. The effect of the PML is that there is no reflection, thus no impedance transformation of the line with some possibly different load impedance. But the line impedance itself remains.

**Reply by nico-arnold** [reply]
Ah, okay. That sounds plausible. Thanks for the clarification!

I'm playing with your TDR model right now, very interesting. Works like a charm. I'll try re-writing this one for testing differential signals inside a CPW and as a small testcase for vias, let's see what comes out there.

**Reply by nico-arnold** [reply]
Small update: I got it working, I think. Thanks a lot for your help, @VolkerMuehlhaus ! Right now it's only single-ended and not differential, but that's a job for tomorrow. 
The model is fully parametric, so you can specify the trace width/length/via diameter/via number etc. and it generates and simulates everything. 
![image](https://github.com/thliebig/openEMS-Project/assets/82146516/eeb54974-3191-4c25-83da-005976ee4170)

Should be close enough for a match, I think. Still, there's a small drop in u2 around 1e-9 s that could be coming from the via - after the wave travelled along the whole line to port2, back to the middle, reflected at the via and then sensed by port2. According to the voltage drop, the via impedance could be slightly lower than the line impedance.
![image](https://github.com/thliebig/openEMS-Project/assets/82146516/9fd16a07-af29-46b0-94f8-642a86c3345d)

 I will try to substitute the vias in the GND planes with boxes tomorrow, maybe that's easier to simulate and also quicker. Plus I can see the raw impact of the via in the middle without the effects of the vias on the GND planes, if any.
I can put the model up here tomorrow morning after a bit of code cleaning if somebody's interested (: Maybe that's a nice usecase to have by hand when designing transmission lines with vias. 

Also, I realized the current travels mostly on the outside of the trace. But that should be normal as the step excitation we use here consists of very high frequencies that, according to the skin effect, are conducted mostly on the outside of the conductor, right?
(SOLVED, see Edit)
![image](https://github.com/thliebig/openEMS-Project/assets/82146516/dadee2ec-2e74-4fca-8efe-e665d4abbb92)

 
 best wishes and good night,
 Nico

Edit: Nevermind for the current density, my bad. I just found out I swapped up the dump_type and file_type...
